src/Hooks.sol

```diff
@@ -15,59 +21,290 @@ contract Hooks is OwnableUpgradeable, IHooks {
     uint256 public constant FEE_DENOMINATOR = 1 ether;
     IVault public immutable VAULT;
     uint256 public immutable VAULT_DECIMALS;
+    Permissions public permissions;
 
+    /**
+     * @notice Constructor
+     * @param vault_ The address of the Vault to which this hooks contract is attached
+     */
     constructor(address vault_) {
         VAULT = IVault(payable(vault_));
         VAULT_DECIMALS = ERC20(address(VAULT)).decimals();
     }
 
-    function initialize(address owner_, uint256 performanceFee_, address performanceFeeRecipient_)
-        external
-        initializer
-    {
+    /**
+     * @notice Initialize the Hooks contract
+     * @param owner_ The address of the owner of the contract
+     * @param performanceFee_ The performance fee to be charged(denominated in 1e18)
+     * @param performanceFeeRecipient_ The address of the performance fee recipient
+     */
+    function initialize(
+        address owner_,
+        uint256 performanceFee_,
+        address performanceFeeRecipient_,
+        Permissions memory permissions_
+    ) external initializer {
         __Ownable_init(owner_);
         if (performanceFee_ > FEE_DENOMINATOR) revert InvalidPerformanceFee();
         if (performanceFeeRecipient_ == address(0)) revert InvalidPerformanceFeeRecipient();
         performanceFee = performanceFee_;
         performanceFeeRecipient = performanceFeeRecipient_;
+        permissions = permissions_;
     }
 
-    function afterProcessAccounting(uint256 totalAssetsBefore, uint256 totalAssetsAfter, uint256 totalShares)
-        external
-    {
+    /**
+     * @notice Modifier to ensure that the caller is the Vault
+     */
+    modifier onlyVault() {
         if (msg.sender != address(VAULT)) revert CallerNotVault();
+        _;
+    }
 
-        if (totalAssetsAfter > totalAssetsBefore) {
-            uint256 yieldEarned = totalAssetsAfter - totalAssetsBefore;
+    /**
+     * @notice After process accounting hook function
+     * @dev This hook is called after the totalBaseBalance is updated
+     * @dev This hook is mints the shares corresponding to the performance fee to the performanceFeeRecipient
+     * @param totalAssetsBeforeAccounting The total assets before accounting
+     * @param totalAssetsAfterAccounting The total assets after accounting
+     * @param totalSupplyBeforeAccounting The total supply before accounting
+     */
+    function afterProcessAccounting(
+        uint256 totalAssetsBeforeAccounting,
+        uint256 totalAssetsAfterAccounting,
+        uint256 totalSupplyBeforeAccounting,
+        uint256,
+        /**
+         * totalSupplyAfterAccounting*
+         */
+        uint256,
+        /**
+         * totalBaseBalanceAfterAccounting*
+         */
+        uint256
+    )
+        /**
+         * totalBaseBalanceBeforeAccounting*
+         */
+        external
+        onlyVault
```

Comment:
i propose we decouple the caller restriction from the vault

so we have onlyOwner

so that we can compose hooks in sequence with a Hook composer down the line.

so for example i can add a hook that adds some invariant checks